### ISSUE TITLE
testng-results.xml was not completely generated.

### ISSUE THREAD

**Issue by ReinhardXu2007** [cbeust/testng#1218]
### TestNG Version

6.9.13.6
Java
1.8.0_74 linux x64
### Actual behavior

As we run tests on Jenkins, and display test report by TestNG jenkins plugin.
After update version of TestNG to 6.9.13.6, and version of java to 1.8.0_74, there are errors occured.

Logs of TestNG jenkins plugin:

> TestNG Reports Processing: START
> Looking for TestNG results report in workspace using pattern: **/testng-results.xml
> Saving reports...
> Processing '/var/jenkins_home/jobs/api-test_payment_app-payment_0001_brood-qa13_daily/builds/679/testng/testng-results.xml'
> Failed to parse XML: CDATA section started on line 1302 and column 26 was not closed (position: TEXT seen ...<value>\n                <![CDATA[org.testng.TestRun... @1302:44) caused by: java.io.EOFException: no more data available - expected end tags </value></param></params></test-method></class></test></suite></testng-results> to close start tag <value> from line 1301 and start tag <param> from line 1300 and start tag <params> from line 1299 and start tag <test-method> from line 1298 and start tag <class> from line 1289 and start tag <test> from line 8 and start tag <suite> from line 5 and start tag <testng-results> from line 2, parser stopped on TEXT seen ...<value>\n                <![CDATA[org.testng.TestRun... @1302:44
> org.xmlpull.v1.XmlPullParserException: CDATA section started on line 1302 and column 26 was not closed (position: TEXT seen ...<value>\n                <![CDATA[org.testng.TestRun... @1302:44) caused by: java.io.EOFException: no more data available - expected end tags </value></param></params></test-method></class></test></suite></testng-results> to close start tag <value> from line 1301 and start tag <param> from line 1300 and start tag <params> from line 1299 and start tag <test-method> from line 1298 and start tag <class> from line 1289 and start tag <test> from line 8 and start tag <suite> from line 5 and start tag <testng-results> from line 2, parser stopped on TEXT seen ...<value>\n                <![CDATA[org.testng.TestRun... @1302:44

And open the testng-results.xml, the file is interrupted.
In some jobs, the result file is generated completely.
In some other jobs, the result file is interrupted on anywhere. It may end with incomplete CDATA "TestRunner", or end with a "&lt;/class&gt;" tag.
### Diagnosis

I created a help class to test which part of logic is wrong.

```
public class XmlReportDiagnosis implements IReporter {

    @Override
    public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites, String outputDirectory) {
        Logger.info(this.getClass(), "generateReport");
        TestNG testNG = TestNG.getDefault();
        if (!CollectionUtils.isEmpty(testNG.getReporters())) {
            for (Iterator<IReporter> iterator = testNG.getReporters().iterator(); iterator.hasNext();) {
                IReporter reporter = iterator.next();
                if (reporter instanceof XMLReporter) {
                    XMLReporter xmlReporter = (XMLReporter) reporter;
                    try {
                        Field rootBufferField = reporter.getClass().getDeclaredField("rootBuffer");
                        rootBufferField.setAccessible(true);
                        XMLStringBuffer rootBuffer = (XMLStringBuffer) rootBufferField.get(xmlReporter);
                        if (null != rootBuffer) {
                            Logger.info(this.getClass(), rootBuffer.toXML());
                        }
                    } catch (NoSuchFieldException e) {
                        e.printStackTrace();
                    } catch (IllegalAccessException e) {
                        e.printStackTrace();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    break;
                }
            }
        }
    }
}
```

In this class, it would print the rootBuffer, a field of XmlReport, to have a preview about whether the content in memory was generated completely.
The test result is, the content in memory is correct.

So the cause was located to 

> Utils.writeUtf8File(config.getOutputDirectory(), FILE_NAME, rootBuffer, null /\* no prefix */);

The last line of XmlReport.generateReport.
I created a new Reporter and replace the output logic with lib apache-commons-lang. It works now.

**Comment by juherr** [comment]
Thanks for the report.

If I understand you well, the issue is located into `Utils.writeUtf8File`?
I suppose we reach the IOException https://github.com/cbeust/testng/blob/master/src/main/java/org/testng/internal/Utils.java#L116
Are you able to reproduce the issue with `Utils.writeUtf8File` only? Or share the expected report produced with commons-lang?

**Comment by ReinhardXu2007** [comment]
Yes, the issue is located into Utils.writeUtf8File.
What weird is, no IOException was thrown, inferred from the log.
And I've tried to use `Utils.writeUtf8File` only.
```
    public static void main(String[] args) {
        XMLStringBuffer buffer = new XMLStringBuffer();
        buffer.push(XMLReporterConfig.TAG_TESTNG_RESULTS);
        for (int i = 0; i < 100000; i++) {
            buffer.push("test", "count", i + "");
            buffer.pop();
        }

        buffer.pop();
        Utils.writeUtf8File(".", XMLReporter.FILE_NAME, buffer, null);
    }
```
It works fine. All the content was written to file.

The key of my fix is just replace this part in `Utils.writeUtf8File` 
```
      try (final OutputStreamWriter w = new OutputStreamWriter(new FileOutputStream(file), "UTF-8")) {
        if (prefix != null) {
          w.append(prefix);
        }
        xsb.toWriter(w);
      }
```
to
```
            String content = xsb.toXML();
            org.apache.commons.io.FileUtils.writeStringToFile(file, content, "UTF-8");
```

**Comment by cantorio** [comment]
Have the same issue after updating from 6.9.10 to 6.9.13.6

**Comment by juherr** [comment]
@cantorio Do you have a simple sample which will help me to reproduce the issue?

**Comment by cantorio** [comment]
Unfortunately, no. All I can say is we have 5500+ tests running via maven. All results are present in testng-results.xml, except last four lines. So the common structure of file is following: 
```xml
<testng-results>
	<suite name="SomeSuite">
		<test name="SomeTest">
			<class name="SomeClass">
				<test-method status="PASS">
					<params>
						<param index="0">
							<value>
								<![CDATA[public %TESTNAME%]]>
							</value>
						</param>
					</params>
					<reporter-output>
					</reporter-output>
				</test-method> <!-- This is the last line, present in testng-results.xml -->
			</class> <!-- Missing line -->
		</test>	<!-- Missing line -->
	</suite> <!-- Missing line -->
</testng-results> <!-- Missing line -->
```
I tried to reproduce the issue with one simple test, but everything was OK.

**Comment by juherr** [comment]
As pointed by @ReinhardXu2007 , it looks it is working with commons-utils but the code is really short: http://grepcode.com/file/repo1.maven.org/maven2/commons-io/commons-io/2.4/org/apache/commons/io/IOUtils.java#IOUtils.write%28java.lang.String%2Cjava.io.OutputStream%2Cjava.nio.charset.Charset%29

On TestNG side, the logic is located in https://github.com/cbeust/testng/blob/070710dbd7dc658d1cba092dd2043292859ced09/src/main/java/org/testng/reporters/FileStringBuffer.java#L69-L88

**Comment by cantorio** [comment]
OK, thanks. As i can see, the issue was fixed in the master branch by adding fw.flush():
https://github.com/cbeust/testng/blob/master/src/main/java/org/testng/reporters/FileStringBuffer.java#L83
Any chance this change will be in maven repo soon?

**Comment by juherr** [comment]
@cantorio Will be available in 6.10 (already on jcenter, maven central soon)

**Comment by juherr** [comment]
Duplicate of #1203

**Comment by cantorio** [comment]
@juherr Thanks!

**Comment by cbeust** [comment]
It's on Maven Central:
http://repo1.maven.org/maven2/org/testng/testng/6.10/

-- 
Cédric

**Comment by Tibor17** [comment]
@juherr 
I have received a bug https://issues.apache.org/jira/browse/SUREFIRE-875.
The problem is that two JVMs are running with TestNG in parallel and one overrides `testng-results.xml` to each other. How can we rename this file on TestNG instance?

**Comment by krmahadevan** [comment]
@Tibor17 - Since they are two different JVMs perhaps the output folder location for each of those JVMs could be controlled via the [configurations](http://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#reportsDirectory) in surefire ? Would that help? 

> How can we rename this file on TestNG instance?

To the best of my knowledge, currently there's no way in which you can do this.

@juherr - Should we go ahead and enable this ? We already support this sort of controlling the naming for Emailable report. See [here](https://github.com/cbeust/testng/blob/master/src/main/java/org/testng/reporters/EmailableReporter.java#L85-L92) and [here](https://github.com/cbeust/testng/blob/master/src/main/java/org/testng/reporters/EmailableReporter2.java#L74-L81)

**Comment by Tibor17** [comment]
@krmahadevan 
You'r right we have to store entire report in an individual subfolder with a test or suite xml or jvm#, e.g.
`target/surefire-reports/TEST-testng.SkipExceptionReportTest/`

**Comment by juherr** [comment]
@Tibor17 I think the @krmahadevan 's workaround is good. 

@krmahadevan Yes, I think we may add support of `JVM_ARG` in all reporters.
I was thinking to centralize the file name logic near `TestNG` which can be, in fact, the instance name.
This new logic may store an index of `TestNG` instances in order to generate an appropriate name: `testng-x` (default name `testng` or `testng-1` or `testng-0`). 
We will be able to use the instance name as thread name too.

**Comment by krmahadevan** [comment]
@juherr 

> This new logic may store an index of TestNG instances in order to generate an appropriate name: testng-x (default name testng or testng-1 or testng-0).

Wouldn't this cause backward compatibility issues ? I think people bank on the file name to do some house cleaning or data parsing from this files etc., 

> Yes, I think we may add support of JVM_ARG in all reporters

I think the XMLReporter is the only one that right now doesnt have this. I have raised a small PR https://github.com/cbeust/testng/pull/1460 which takes care of this as well.

**Comment by juherr** [comment]
@krmahadevan 

> Wouldn't this cause backward compatibility issues? 

I don't think so because if the default name is `testng` and the reporter add its own logic on name (`-results.xml`), it will be `testng-results.xml` as usual.

**Comment by eliasbalasis** [comment]
Coming from original https://issues.apache.org/jira/browse/SUREFIRE-875,

I am glad you agree with the "individual subfolder" approach.
For consistency that is what I did as part of a TestNG listener for my problem (testng-results.xml in subfolder)

It would be nice if TestNG itself could handle this transparently.

**Comment by juherr** [comment]
@eliasbalasis I think the merge of suite files should be located in surefire but we can help to do it if @Tibor17 likes the idea :)

**Comment by eliasbalasis** [comment]
Thanks @juherr,

I agree, consolidation of test results (consolidated testng-results.xml in standard location) is not a task for TestNG itself, not to say it is not really something we cannot live without :-)

However, the separation of test results in a different file per TestNG execution (preferably in individual sub-folder) is very much required and is a task for TestNG.

Question 1
How can TestNG be instructed to generate test results in different files?
An additional TestNG parameter perhaps?

Question 2
How can Surefire be amended to pass this extra parameter to its TestNG executions?
Perhaps a question for Surefire team?

**Comment by juherr** [comment]
Answer 1:
In the next TestNG release, `testng.report.xml.name` env var will be available to specify the custom file name.

**Comment by eliasbalasis** [comment]
Shouldn't this be a parameter instead of an env variable? so that Surefire caller can direct TestNG to generate results in a different file?

**Comment by Tibor17** [comment]
@juherr 
The outcome of TestNG execution including these report xml files is configured via `TestNG.setOutputDirectory()`?

>In the next TestNG release, testng.report.xml.name env var will be available to specify the custom file name.
Env vars and global configuration is inherently bad idea because people may not expect it when all Jenkins jobs are misleadingly influenced.

@krmahadevan 
maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#reportsDirectory
is NOT only for TestNG.
Moreless the folder name should be dynamically dicovered. I don't think that user can list all jvm# folders in configuration. What if the number of forks does not match the number of folders he configured?
Therefore Surefire should discover some kind of generic approach.

**Comment by eliasbalasis** [comment]
I agree with the dynamic discovery part. Surefire should direct TestNG (or other).
Shouldn't dynamic discovery be controlled by additional Surefire parameter?

Is TestNG.setOutputDirectory() a valid entry point that could be used by Surefire to direct TestNG?

**Comment by Tibor17** [comment]
@eliasbalasis 
We call it.
My proposal is to still use target/surefire-reports if JVM is only one or inplugin process.
If more jms then e.g. 2 jvms
target/surefire-reports/testng-1
target/surefire-reports/testng-2
etc
This should not break backwards compatibility. In case of one JVM absolutely not and in case of 2 or more the old reports of two JVMs have not been ever merged together so it means useless. Therefore I think nobody who needed testng xml report used `forkCount` or `forkMode` in Surefire.

**Comment by eliasbalasis** [comment]
I agree.

Most people are using a single TestNG suite and no forks, so I bet not many people have experienced this and that's perhaps why it has escaped attention.

On a single JVM or single TestNG suite scenario, of course, no need for dedicated sub-folder.

If JVMs >= 2 or TestNG suites >= 2 (forked or not) then a dedicated sub-folder (working folder) is required.

**Comment by Tibor17** [comment]
> However, this is all Surefire territory. Shall we escalate this to Surefire team?
No it is not because TestNG create reports and the responsibility should be in there. If report content changes, for any major change or minor one, the TestNG should then provide again consistent merge report.
The report does not only have this XML file. There is entire HTML. Do you want to merge all these?

**Comment by eliasbalasis** [comment]
I see,
I thought TestNG generates output in folder Surefire runs it under, Surefire controlled.

I am hoping for
target/surefire-reports/exec1/...
target/surefire-reports/exec2/...
etc.

Similarly for failsafe plugin

Makes sense?

**Comment by Tibor17** [comment]
@eliasbalasis
Surefire creates own report, i.e. `TEST-TestSuite.xml` and `TestSuite.txt`, but the rest belongs to TestNG.
I think merging parallel TestNG suites would be complex task for TestNG team as well.
Therefore multiple folder is easy on Surefire side.

`cd target/surefire-reports/`

06. 07. 2017  18:24    <DIR>          .
06. 07. 2017  18:24    <DIR>          ..
06. 07. 2017  18:24             3 059 emailable-report.html
06. 07. 2017  18:24               734 index.html
06. 07. 2017  18:24    <DIR>          Surefire suite
06. 07. 2017  18:24             8 715 TEST-TestSuite.xml
06. 07. 2017  18:24               898 testng-results.xml
06. 07. 2017  18:24               312 testng.css
06. 07. 2017  18:24               271 TestSuite.txt
               6 File(s)         13 989 bytes
               3 Dir(s)  197 462 384 640 bytes free

`cd "Surefire suite"`

06. 07. 2017  18:24    <DIR>          .
06. 07. 2017  18:24    <DIR>          ..
06. 07. 2017  18:24             1 340 classes.html
06. 07. 2017  18:24                38 groups.html
06. 07. 2017  18:24               187 index.html
06. 07. 2017  18:24               120 main.html
06. 07. 2017  18:24             1 185 methods-alphabetical.html
06. 07. 2017  18:24                50 methods-not-run.html
06. 07. 2017  18:24             1 185 methods.html
06. 07. 2017  18:24                39 reporter-output.html
06. 07. 2017  18:24             2 053 Surefire test.html
06. 07. 2017  18:24                27 Surefire test.properties
06. 07. 2017  18:24               354 Surefire test.xml
06. 07. 2017  18:24               820 testng.xml.html
06. 07. 2017  18:24             1 181 toc.html
              13 File(s)          8 579 bytes
               2 Dir(s)  197 462 384 640 bytes free

**Comment by eliasbalasis** [comment]
>> Therefore multiple folders is easy on Surefire side.

I cannot agree more on that.

Shall we treat this as a Surefire related issue then?

Any thoughts?

**Comment by juherr** [comment]
@Tibor17 @eliasbalasis Please, could you summarize what the TestNG team is supposed to do in a new issue?

**Comment by eliasbalasis** [comment]
Well, to my understanding, there is nothing to do on TestNG side.

I believe it is Surefire/Failsafe related.

If @Tibor17 agrees I am happy to close and escalate this to Surefire/Failsafe team.

@Tibor17 any thoughts?

**Comment by Tibor17** [comment]
@eliasbalasis
I am Apache team ;-)
Do you want to be volunteer and open pull request on GitHub?

**Comment by eliasbalasis** [comment]
So, we agree that this is not a TestNG issue but a Surefire/Failsafe issue.

@juherr I believe you can close this.

@Tibor17 I will continue in Surefire issue tracker. I will try to open a pull request.

Thanks guys

**Comment by Tibor17** [comment]
@eliasbalasis
We want to override HEAD commit in maven-surefire.
If you can, please wait with PR and meanwhile prepare your changes in working copy yet.

**Comment by nzin-appdirect** [comment]
Hi guys,
not sure it is the best place to ask that, but: I just faced the problem you are mentioning (my maven-failsafe-plugin is configured to use several fork (<forkCount>2</forkCount>)), and testng-results.xml is overwritten by each thread/fork.
I tried to configure the reportsDirectory in my pom.xml, but so far I didn't manage to set it properly. I tried something like:
```
<plugin>
<artifactId>maven-failsafe-plugin</artifactId>
<reportsDirectory>${project.build.directory}/surefire-reports/${surefire.forkNumber}</reportsDirectory>
...
</plugin>
```
But I finish with a directory named "target/surefire-reports/${surefire.forkNumber}/testng-results.xml"

i.e. the variable is not interpolated
Do you known which value I can use instead of surefire.forkNumber?

**Comment by juherr** [comment]
You should try to ask your question on stackoverflow or maybe on the surefire issue tracker.
